### ISSUE TITLE
Track NPCs through the networking layer

### ISSUE THREAD

**Issue by essial** [OpenDiablo2/OpenDiablo2#350]
Currently NPCs are handled on the map engine directly. This needs to be moved to the new networking infrastructure so that they sync up for all players on the game.

**Comment by danhale-git** [comment]
I'll take this.

**Comment by danhale-git** [comment]
I had a look at this tonight, found that NPCs satisfy d2mapentity.MapEntity and are stored in d2mapengine.MapEngine.entities. Presumably MapEntity will include monsters also. I have read through d2client and d2server.

@essial does this sound correct:

- I'll need to change (add to) d2client.OnPacketRecieved, d2server.runNetworkServer and d2server.OnClientConnected (at least).
- I should sync entity movement, basing my code on the current player movement.
- I should sync entity create/remove.

I have a bit more reading to do, notably map generation. Any direction you can provide is appreciated. I'll probably work on this next tomorrow, when you're offline.

**Comment by danhale-git** [comment]
> Solution assumes that, if all map entities are synchronised on connection, they will stay synchronised. Is that the case?
It is not the case. That was not a viable solution. Useful discussion [here](https://discord.com/channels/515518620034662421/515518620512681988/725712378863288331).

**Comment by essial** [comment]
Yup all movement commands should be sent by the server and not handled automatically by the map engine or client. Additionally, although the original D2 didn't track things like chickens, I think the server in our engine should handle ALL entity movement. There is no reason to optimize here IMO and would actually cause things to be a bit more convoluted.

**Comment by danhale-git** [comment]
I have been unable to find a satisfactory solution for this yet, mainly due to a lack of knowledge about the project I think. The issue has become 'find a sensible way to track all entities though the networking layer' which leads do 'maybe refactor entity types' at which point I'm a little out of my depth. 

I think this issue might be helped by a little refactoring of the various types which declare `mapEntity` (NPC, Object, AnimatedEntity, Player).

I'm going to spend some time reading code and probably fix some lint issues while I'm there. I will hold on to this issue and comment here when I am working on it again (soon probably).

If anyone else feels the need to bang this issue out in the mean time, I will be very interested to see how you do it.
